docs/design/git-submodule-storage.md

```diff
@@ -0,0 +1,138 @@
+# Git submodule storage
+
+## Objective
+
+Decide what approach(es) to Git submodule storage we should pursue.
+The decision will be recorded in [./git-submodules.md](./git-submodules.md).
+
+## Use cases to consider
+
+The submodule storage format should support the workflows specified in the
+[submodules roadmap](./git-submodules.md). It should be obvious how "Phase 1"
+requirements will be supported, and we should have an idea of how "Phases 2,3,X"
+might be supported.
+
+Notable use cases and workflows are noted below.
+
+### Fetching submodule commits
+
+Git's protocol is designed for communicating between copies of the same
+repository. Notably, a Git fetch calculates the list of required objects by
+performing reachability checks between the refs on the local and the remote
+side. We should expect that this will only work well if the submodule repository
+is stored as a local Git repository.
+
+Rolling our own Git fetch is too complex to be worth the effort.
+
+### "jj op restore" and operation log format
+
+We want `jj op restore` in the superproject to restore to an "expected" state in
+the submodule, though the definition of "expected" is up for debate. At the very
+least, `jj op restore` should restore the submodule working copy, but we may
+want to support the full list of operations at some point.
+
+Currently, the operation log only references objects and refs in the
+superproject, so it is likely that proposed approaches will need to extend this
+format. It is also worth considering that submodules may be added, updated or
+removed in superproject commits, thus the list of submodules is likely to change
+over the repository's lifetime.
+
+### Nested submodules
+
+Git submodules may contain submodules themselves, so our chosen storage schemes
+should support that.
+
+### Bonus: supporting future extensions
+
+There are certain extension we may want to make in the future, but are non-goals
+today. When evaluating approaches, consider these as low-priority "bonus"
+points; they shouldn't weight heavily in our decision-making, but all else being
+equal, we should prefer the approach that is more easily extended.
+
+These extensions are:
+
+- Non-git subrepos
+- Colocated Git repos
+
+## Possible approaches
+
+### Approach 1: Git repos in the main Git backend
+
+Since the Git backend contains a Git repository, an 'obvious' default would be
+to store them in the Git superproject the same way Git does, i.e. in
+`.git/modules`. This should allow us to _mostly_ reuse the existing backend
```

Comment:
I think this suggests that we want our solution to be generalizable to the various jj use cases. I'll probably expand the criteria to capture this, probably under `Simplicity`.

Reply:
This sounds like "non-git backends" are a possible extension under `### Bonus: supporting future extensions`, and that extensibility is actually a major concern instead of `low-priority "bonus" points`. Should I reword that section? Or maybe this will naturally show up when we evaluate proposals based on their simplicity.

Reply:
Yes, to the extent that it affects design, I think it's a major concern.

Reply:
Okay, that's enough reason to reject this idea. I'll send a quick update to that effect.

Reply:
Actually, this is enough reason to reject `Store Git submodules as alternate Git backends` too, so I think this criterion is enough for us to converge on just 1.5-ish ideas. I'll overhaul this doc to present the ideas correctly.

Reply:
Sounds good. Thank you